### PR TITLE
feat: support openai structured output

### LINKED ISSUE #1197: [Feature Request] Support lists of BaseModels in StructuredOutputs
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

A very common use case is that we will want to create a list of things, but do not have a type specifically for a list as list[thing] already serves well for that.

Right now if you attempt to ChatAgent.step with a response format of list[BaseModel], it will show a type error `Expected type 'Type[BaseModel] | None', got 'Type[list[BaseModel]]'` instead  and at runtime throw`AttributeError: type object 'list' has no attribute 'model_json_schema'`

### Solution

Support list[BaseModel] where BaseModel may be any pydantic type in structured outputs.

## Constrained types
While we're here, it can be very valuable to also support [Constrained types](https://docs.pydantic.dev/1.10/usage/types/#constrained-types), particularly constrained lists. We can add steps to ensure the constraints are met, like writing in the description field for the list of the wrapping type the number of elements expected, and adding a wrapper around the element that includes a counter to allow the LLM to keep track of how many elements its made, which is thrown away when returned.

This will be very useful for ensuring a minimum and maximum number of elements are generated, which will be very useful for generating data.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Wendong-Fan**
## Description

support openai structured output

## Motivation and Context

`close #1197 ` 

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/models/openai_model.py [hunk 1]
```diff
@@ -127,13 +128,49 @@ def run(
             self.model_config_dict["presence_penalty"] = 0.0
             self.model_config_dict["frequency_penalty"] = 0.0
 
+        if self.model_config_dict.get("response_format"):
+            # stream is not supported in beta.chat.completions.parse
+            if "stream" in self.model_config_dict:
+                del self.model_config_dict[key]
```
Comment (CaelumF) [inline camel/models/openai_model.py:134]:
```suggestion
                del self.model_config_dict["stream"]
```

File: camel/models/openai_model.py [hunk 2]
```diff
@@ -127,13 +128,49 @@ def run(
...
             messages=messages,
             model=self.model_type,
             **self.model_config_dict,
         )
         return response
 
+    def _to_chat_completion(
+        self, response: "ParsedChatCompletion"
+    ) -> ChatCompletion:
+        choice = dict(
```
Comment (CaelumF) [inline camel/models/openai_model.py:154]:
```suggestion
        # TODO: Handle n > 1 or warn consumers it's not supported
        choice = dict(
```

**Review by CaelumF — CHANGES_REQUESTED** [review]

**Review by lightaime — COMMENTED** [review]
"When using the SDK, you can use the parsed attribute to access the parsed JSON response as an object. This object will be of the type you defined in the response_format parameter."

```
math_response = completion.choices[0].message.parsed
print(math_response.steps)
print(math_response.final_answer)
```

This feature is very good.Any plan to add it? Maybe we should like to into:

https://github.com/openai/openai-python

to see how it is implemented

**Comment by lightaime** [comment]
https://platform.openai.com/docs/guides/structured-outputs

**Pull request merged**